### PR TITLE
Elements: Add unit tests to cover both classname application and style generation

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
This is a backport PR that includes the following PHP Gutenberg changes:

https://github.com/WordPress/gutenberg/pull/55113

It provides additional unit tests for the elements block support and its associated classname and style generation.

Trac ticket: https://core.trac.wordpress.org/ticket/59555

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/block-supports/elements.php [hunk 1]
```diff
@@ -3,171 +3,306 @@
  * @group block-supports
  *
  * @covers ::wp_render_elements_support
+ * @covers ::wp_render_elements_support_styles
```
Comment (costdev) [inline tests/phpunit/tests/block-supports/elements.php]: Test classes should only have `@covers` annotations if all tests cover all of the functions listed in the annotation.

There's been an effort recently to follow a pattern of separating tests for a given function into their own file, so I'd recommend putting the `wp_render_elements_support_styles()` tests in their own file and class:

File name: `wpRenderElementsSupportStyles.php`
Class name: `Tests_Block_Supports_WpRenderElementsSupportStyles`
Covers annotation: `@covers ::wp_render_elements_support_styles`

File: tests/phpunit/tests/block-supports/elements.php [hunk 2]
```diff
@@ -3,171 +3,306 @@
...
+	 * @var string|null
 	 */
-	private static function make_unique_id_one( $block_content ) {
-		return preg_replace( '/wp-elements-[a-zA-Z0-9]+/', 'wp-elements-1', $block_content );
+	private $test_block_name;
+
+	public function set_up() {
+		parent::set_up();
+		$this->test_block_name = null;
+	}
```
Comment (costdev) [inline tests/phpunit/tests/block-supports/elements.php]: This fixture method shouldn't be needed as `$this->test_block_name` is already reset in `::tear_down()`.

**Review by costdev — COMMENTED** [review]
Thanks for the PR @aaronrobertshaw! I've left some feedback above :slightly_smiling_face:

File: tests/phpunit/tests/block-supports/elements.php [hunk 3]
```diff
@@ -3,171 +3,306 @@
...
+				'supports'    => array(
+					'color' => $color_settings,
+				),
 			)
 		);
-		$this->assertSame(
-			$result,
-			'<p class="wp-elements-1">Hello <a href="http://www.wordpress.org/">WordPress</a>!</p>'
-		);
 	}
```
Comment (felixarntz) [inline tests/phpunit/tests/block-supports/elements.php]: I think this logic should run as part of `set_up()` instead of in the individual tests. So either this method should be removed, or alternatively it should be called in `set_up()`. Also, if we keep the method, it should become `private`.

The block is always registered like that, and `tear_down()` calls the equivalent `unregister_block_type()`, so there should be parity with that.

File: tests/phpunit/tests/block-supports/elements.php (hunk 2 above)
Comment (felixarntz) [reply tests/phpunit/tests/block-supports/elements.php]: Per my other comment, I think we should use this fixture to register the block type though instead of doing it in the individual tests.

File: tests/phpunit/tests/block-supports/elements.php (hunk 1 above)
Comment (felixarntz) [reply tests/phpunit/tests/block-supports/elements.php]: +1 for removing the `@covers` annotations from the class and putting on the individual functions.

However, I'm not sure about moving to function-specific files. @costdev Can you point to where this was decided? Or is being discussed? I have never heard about this guidance before, so I want to make sure we're not making changes to the unit tests structure preliminarily.

**Review by felixarntz — COMMENTED** [review]
Thanks @aaronrobertshaw, this looks great so far. I left a bit of additional feedback.

File: tests/phpunit/tests/block-supports/elements.php (hunk 1 above)
Comment (costdev) [reply tests/phpunit/tests/block-supports/elements.php]: @felixarntz It's not that it was decided as a change - There's a bit of a mixed history in the Core test suite of having one file per function, or dumping everything into one place (e.g. option.php for one we've seen recently). So both approaches are currently fine.

The [Writing PHPUnit Tests Handbook entry](https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#test-classes) gives an example of a file that tests the `get_comment_class()` function, with its own file and class. The handbook entry doesn't impose any restrictions, but as this file was previously only testing one function, and will now be testing two, I thought I'd suggest creating a separate file.

The effort I'm referring to is just that several Gutenberg test files have been split into per-function files, and while this divides up a given feature's tests (which would otherwise be put into one file), it isolates a given function's tests, making them easier to find.

Comment (costdev) [reply tests/phpunit/tests/block-supports/elements.php]: Actually, allow me to correct myself:

The handbook entry actually says:
> As a rule, a single test class should not contain tests for more than one function/method.
> [Reference](https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#test-classes)

Noting that it says _should not_, rather than _must not_, so there's room for nuance. Personally, I think of it this way: As this function's tests are all new, there's no real need to place them in this file with another function's tests. There could be any number of additional tests added in future (for these or other functions), and this could become a bit messy and hard to follow/maintain. Giving its tests their own file puts them in the same directory as the other function, but isolated for running and expansion. Some of the test suite also puts shared resources such as helper methods, set_up/tear_down, and data providers in a file (such as `base.php`) to be imported by files that need them. So even if 10 functions find use for these resources, the functions don't all have to be tested in the same file, and instead just import or extend in their own test file.

So yeah, it's my recommendation based on handbook guidance and one pattern visible in the test suite, rather than a "_must_".

Comment (aaronrobertshaw) [reply tests/phpunit/tests/block-supports/elements.php]: Thanks for the feedback 👍 

I've split the tests into separate files per function as per this [comment](https://github.com/WordPress/wordpress-develop/pull/5418#discussion_r1348478534). See https://github.com/WordPress/wordpress-develop/pull/5418/commits/a7dbbb7d5c08775aaab81a414d6977137bfef605

File: tests/phpunit/tests/block-supports/elements.php (hunk 2 above)
Comment (aaronrobertshaw) [reply tests/phpunit/tests/block-supports/elements.php]: Good catch, I've cleaned that up in https://github.com/WordPress/wordpress-develop/pull/5418/commits/a7dbbb7d5c08775aaab81a414d6977137bfef605.

File: tests/phpunit/tests/block-supports/elements.php (hunk 3 above)
Comment (aaronrobertshaw) [reply tests/phpunit/tests/block-supports/elements.php]: I'm likely missing something here but I'm not following the suggestion. The block type's supports config differs across the data sets used by the test.

While splitting these tests into separate files per function, I've moved the block type registration into the test itself so the per-test configuration might be more obvious.

**Comment by aaronrobertshaw** [comment]
Appreciate all the feedback @costdev & @felixarntz 👍 

I believe I've addressed everything now. Let me know if I missed anything

**Review by felixarntz — APPROVED** [review]
Thanks for the updates @aaronrobertshaw, LGTM!

**Review by costdev — APPROVED** [review]
Thanks @aaronrobertshaw! Looks good to go!

**Comment by felixarntz** [comment]
Committed in https://core.trac.wordpress.org/changeset/56806 - which for some reason doesn't show up yet, but proof it was committed can be found here: https://github.com/WordPress/wordpress-develop/commit/4d5859bececc8b451ca95108d942f2a870bdf806 😆
